File: experiments/nexmark/queries/flink/src/main/java/org/example/queries/Q1.java
package org.example.queries;

import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonPropertyOrder;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.example.data.Bid;
import org.example.io.DataSink;

public class Q1 {

    @JsonPropertyOrder({"auction", "price", "bidder", "dateTime"})
    public static class Output {
        public long auction;
        public long price;
        public long bidder;
        public long dateTime;

        public Output(long auction, long price, long bidder, long dateTime) {
            this.auction = auction;
            this.price = price;
            this.bidder = bidder;
            this.dateTime = dateTime;
        }

        public Output() {
        }
    }

    public static void run(DataStream<Bid> bids) {
        bids
                .map(b -> new Output(b.auction, (long) Math.floor(b.price * 0.85), b.bidder, b.dateTime))
                .returns(Output.class)
                .addSink(new DataSink<Output>());
    }

    public static void runOpt(DataStream<Bid> bids) {
        bids
                .map(b -> new Output(b.auction, (long) Math.floor(b.price * 0.85), b.bidder, b.dateTime))
                .returns(Output.class)
                .addSink(new DataSink<Output>());
    }
}

File: experiments/nexmark/queries/shark/src/q1.rs
use runtime::prelude::*;

use crate::data::Bid;

#[data]
struct Output {
    auction: u64,
    price: u64,
    bidder: u64,
    date_time: u64,
}

pub fn run(bids: Stream<Bid>, ctx: &mut Context) {
    bids.map(ctx, |bid| {
        Output::new(bid.auction, bid.price * 100 / 85, bid.bidder, bid.date_time)
    })
    .drain(ctx);
}

pub fn run_opt(bids: Stream<Bid>, ctx: &mut Context) {
    bids.map(ctx, |bid| {
        Output::new(bid.auction, bid.price * 100 / 85, bid.bidder, bid.date_time)
    })
    .drain(ctx);
}
